Add how to run data migration docs (#1798)
Enterprise-CMCS/managed-care-review: docs/technical-design/howto-migrations.md

# How to complete migrations

## Background
Migrations are a controlled way to change existing records in the database or else change the shape of tables/relationships. They are necessary when fields are added, removed, or when the meaning behind fields and their relationships change.

## General Guidance

> - Prioritize testing in environments that have data similar to production. If you are testing on a review app, you should pre-populate data through using the deployed app a user or via Cypress runs to have a diversity of test data available.

Reviewer: ```suggestion
- Prioritize testing in environments that have data similar to production. If you are testing on a review app, you should pre-populate data through using the deployed app as a user or via Cypress runs to have a diversity of test data available.
```

> - Run in all environments, in as close to a way as they run in CI.

Reviewer: Maybe spell this out a bit more? I'm not sure I know what it means.

Author: Oops yes. That is very vague. The long version is I found out that though you could run it from the CLI with `severless` commands which is a lot faster than waiting for PR apps to deploy and merge (which really slows feedback incredibly) @mojotalantikite feels its much better to run through the lambda in AWS because you are running in more of CI-like way in terms of relying on the AWS env. So I think its good to know to prefer that method. I will figure out a couple word way to say that succinctly.

> - Write migrations that can be tested repeatedly.

Reviewer: ```suggestion
- Write migrations that can be tested repeatedly.  Include logic to skip or handle data that has already been changed or doesn't need to be migrated, so that repeated runs of the lambda don't result in bad data.
```

- Avoid making migrations directly dependent on a specific feature in progress being available in production.

## Data Migrations
Currently data migrations change existing records in the database. They are run as standalone lambdas that developers must manually trigger by environment.

### Steps
1. Prepare for manual testing in lower environments.

>     - Build a PR review app off `main` to start out. Try to populate with submissions similar to what is in production (For example, populate submissions via Cypress and there are specific submissions types you know you will need to test, make sure add them).

Reviewer: ```suggestion
    - Build a PR review app off `main` to start out. Try to populate with submissions similar to what is in production (For example, populate submissions via Cypress and if there are specific submissions types you know you will need to test, make sure add them).
```

    - Console statements are essential because there are line numbers in Cloudwatch when the migration fails or does not apply correctly. There is little additional context about why a run execution paused or failed.
    - Here is an example of verbose debug console statements wrapping a loop:
    ![debug consoles in data migration script](../../.images/verbose-logs-example.png)
1. Write unit tests.
    - Consider including a test for `migration can be run repeatedly without data loss or unexpected results". Using repeatable migrations (when possible) makes it easier on developers and easier to unwind work.
1. Manually test the migration in your review app.
    - Log into [AWS Lambda Console](https://console.aws.amazon.com/lambda/home) and find the lambda. This means choosing `app-api-` lambda with your branch name and migration name included.
    - Click the Test tab and the `Test` button. Use the generic hello world event.
        - All output will appear inline on the same page as the lambda run. All consoles appear there as well. You can also click into a link from there into Cloudwatch to be able to see entire output of logs around the lambda execution.

>      - If you need to debug quickly, without waiting for redeploy in review app use the severless CLI script.

Reviewer: If I'm remembering right, this needs to be run from app-api?  If that's correct, probably worth mentioning.

>         -  `serverless deploy function --function name_of_migration --stage my_branch && echo "DEPLOYED" && serverless logs --function migrate_rate_documents  --stage my_branch --tail` can b e run in your `app-api` directory. This script runs serverless off your local API code (configured with your env variables) with the AWS resources associated with your branch.  `my_branch`is the name of your deployed Github branch and `name_of_migration` is the name of the migration file mind the file path.

Reviewer: ```suggestion
        -  `serverless deploy function --function name_of_migration --stage my_branch && echo "DEPLOYED" && serverless logs --function name_of_migration  --stage my_branch --tail` can be run in your `app-api` directory. This script deploys your local handler code directly to AWS, without going through our normal CI build process.  `my_branch`is the name of your deployed Github branch and `name_of_migration` is the name of the migration file mind the file path.
```

Reviewer: Thought this was more clear, but feel free to keep or change what you had.

>         - If this is to be run in DEV (`main`) you will have to reshape your `envrc.local` to imitate configuration on that stage.

Reviewer: This might not make a lot of sense to someone who doesn't already know what it means. Maybe include a screenshot of a lambda config?  Let me see if this paste works.
<img width="1598" alt="image" src="https://github.com/Enterprise-CMCS/managed-care-review/assets/8964335/befc9607-c3d2-4c02-b75c-eef0a5e8f2db">

>         - Command line serverless scripts should not be used on higher environments than dev - it's important to run the migration via AWS lambda directly since that configuration is more similar to how our promote CI works.

Author: +1 thank you. Similar comment to https://github.com/Enterprise-CMCS/managed-care-review/pull/1798#discussion_r1258488218 I will look into the wording I'm using it is not clear.

1. Define clear acceptance criteria, this will be re-used.

>     - How do you know migration applied? What will be checked either in the app, via api request, or in the reports CSV?

Reviewer: love this
